### ISSUE TITLE
How it is predicting that a person is male or a female?

### ISSUE THREAD

**Issue by pallavipundir** [clintval/gender-predictor#5]
As I have a name "hgo" in my file, it can be male or a female. So how it will be decided by the code to predict it as a male or a female?

**Comment by clintval** [comment]
Hi @pallavipundir this project aims to predict the gender (male/femaale) of a person based on the top American baby names in 2015 with at least five occurrences.

The classifier is [Naive Bayes](https://www.wikiwand.com/en/Naive_Bayes_classifier) which learns based on the training data.

Names that fit both genders well are difficult to classify so there is more error on names like that.

**Comment by pallavipundir** [comment]
Hi Clint, Thank you for your response.I totally agree with you.I am stuck somewhere in the code, I just want to know about your male probability logic.What exactly the code is doing with male probability.This line:
 m_prob = 0.01 if m_prob == 0 else 0.99 if m_prob == 1 else m_prob
What's the use of this line?How can we predict a gender with this line?

**Comment by clintval** [comment]
That line prevents probabilities from being absolutely 0% or 100% which are logically (and philosophically) impossible as well as cause division by zero errors.

The creation of the classifier occurs here:

https://github.com/clintval/gender-predictor/blob/7eda5f5d611d49fe0dc3312f0bbe90aabae7065a/gender_predictor/__init__.py#L42-L48

And that classifer is used to classify a new name here:

https://github.com/clintval/gender-predictor/blob/7eda5f5d611d49fe0dc3312f0bbe90aabae7065a/gender_predictor/__init__.py#L39-L40

This is, admittedly, an old project for me and maybe not as extensible as it should be. Let me know if you need additional help. I might be able to provide it.

**Comment by pallavipundir** [comment]
Also I want to know that Is there any method defined in your code that predict the gender of only one kind of those names that fit both genders well.

**Comment by clintval** [comment]
Ah, that's a good question. You can use the classifier but you will have to call an additional method.

Something like:

```python
from gender_predictor import GenderPredictor

name = 'hgo'

gp = GenderPredictor()
gp.train_and_test()
gp.classifier.prob_classify(gp._name_features(name.upper()))
```

I should not that the name `"Hgo"` is not a common American name so the default training set for this library will likely not classify your names well!

**Comment by pallavipundir** [comment]
Yeah sure.Thank you Clint.But I am stuck here like suppose I have a name in my file that is "J" and "hgo" .And your code is predicting both of them as a "Male".How?Could you please define it?

**Comment by clintval** [comment]
@pallavipundir The code that performs the classify is the NLTK Naive Bayes classifier.

You can read about that here: https://www.nltk.org/_modules/nltk/classify/naivebayes.html

The features that I am feeding into the classifier are listed here as there is literature to support these as being important features in an American name to guess gender:

https://github.com/clintval/gender-predictor/blob/7eda5f5d611d49fe0dc3312f0bbe90aabae7065a/gender_predictor/__init__.py#L52-L57

**Comment by pallavipundir** [comment]
Thank you so much Clint.

**Comment by clintval** [comment]
No problem, and good luck!

Like I said, this library is pretty old and I was just learning Python.

One thing it does not easily support is new training data so let me know if you would like to work together on something new.

**Comment by pallavipundir** [comment]
Yeah sure.I would love to.I'll let you know for sure.

**Comment by pallavipundir** [comment]
Hi Clint,I have a new question.Can I add more name with gender files in names.zip to increase the accuracy?

**Comment by pallavipundir** [comment]
Hi Clint,I have a new file.Can we both work on that?
